File: cpp/minimum_path_sum.cpp
64. Minimum Path Sum Add to List
DescriptionHintsSubmissionsSolutions
Total Accepted: 105901
Total Submissions: 280339
Difficulty: Medium
Contributor: LeetCode
Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

Subscribe to see which companies asked this question.

Hide Tags Array Dynamic Programming
Hide Similar Problems (M) Unique Paths (H) Dungeon Game

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m=grid.size();
        if (m==0) return 0;
        int n=grid[0].size();
        if (n==0) return 0;
        vector<vector<int>> dp(m, vector<int>(n, 0));
        
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (i==0&&j==0) {
                    dp[i][j]=grid[i][j];
                } else {
                    dp[i][j]=grid[i][j] + min(i>0?dp[i-1][j]:INT_MAX, j>0?dp[i][j-1]:INT_MAX);    
                }
            }
        }
        return dp[m-1][n-1];
    }
};

2. Use 1D dp array
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m=grid.size();
        if (m==0) return 0;
        int n=grid[0].size();
        if (n==0) return 0;
        vector<int> dp(n, 0);
        
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (i==0&&j==0) {
                    dp[j]=grid[i][j];
                } else {
                    dp[j]=grid[i][j] + min(i>0?dp[j]:INT_MAX, j>0?dp[j-1]:INT_MAX);    
                }
            }
        }
        return dp[n-1];
    }
};

File: java/minimum_path_sum.java
public class Solution {
    public int minPathSum(int[][] grid) {
        int m=grid.length;
        if (m==0) return 0;
        int n=grid[0].length;
        if (n==0) return 0;
        
        int[][] dp=new int[m][n];
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (i==0 && j==0) dp[i][j]=grid[i][j];
                else if (i==0) dp[i][j]=dp[i][j-1]+grid[i][j];
                else if (j==0) dp[i][j]=dp[i-1][j]+grid[i][j];
                else {
                    dp[i][j]=Math.min(dp[i][j-1],dp[i-1][j])+grid[i][j];
                }
            }
        }
        
        return dp[m-1][n-1];
    }
}

2.
public class Solution {
    public int minPathSum(int[][] grid) {
        int m=grid.length;
        if (m==0) return 0;
        int n=grid[0].length;
        if (n==0) return 0;
        
        int[] dp=new int[n];
        dp[0]=grid[0][0];
        for (int j=1; j<n; j++) {
            dp[j]=dp[j-1]+grid[0][j];
        }
        
        for (int i=1; i<m; i++) {
            dp[0]+=grid[i][0];
            for (int j=1; j<n; j++) {
                dp[j]=Math.min(dp[j-1],dp[j])+grid[i][j];
            }
        }
        
        return dp[n-1];
    }
}
